# Exploring Ebay Car Sales Data

---

## 1.  Introduction


This project is created to demonstrate the basics of data cleaning and data exploration using *pandas*.

The dataset used in this project contain records of used cars from eBay-Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data), but a few modifications have been made by DataQuest to the original dataset. About 50,000 data points were sampled from the full dataset, and the data was dirtied to make it more closely resemble what would be expected from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with).

The data dictionary provided with data is as follows:
- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer.
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **kilometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.

   Note: The fields **lastSeenOnline** and **dateCreated** could be used to estimate how long a car will be at least online before it is sold.


## 2.  Goal & Objective of This Project

This project aims to clean the data and analyze the included used car listings.








## 3. General Observation of the Dataset

In [1]:
### Import the pandas and NumPy libraries.
### Reading the dataset into pandas.

import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

### Note: Common encoding is "UTF-8" / "Latin-1" / "Windows-1252".

In [2]:
### Showing a preview of the dataset.

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
### Showing the first 5 rows of data.

autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
### Getting the overview of all the dtypes used in the dataset, along with its shape, columns, and other information.

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
### Check columns that have null values.

check_null = autos.isnull().sum()
check_null

dateCrawled               0
name                      0
seller                    0
offerType                 0
price                     0
abtest                    0
vehicleType            5095
yearOfRegistration        0
gearbox                2680
powerPS                   0
model                  2758
odometer                  0
monthOfRegistration       0
fuelType               4482
brand                     0
notRepairedDamage      9829
dateCreated               0
nrOfPictures              0
postalCode                0
lastSeen                  0
dtype: int64

**Observations from the above:**

- The dataset contains 20 columns, most of which are strings (objects).

- Some columns have null values, but none have more than ~20% null values.

- Most of the dates columns are stored as strings (objects).

- The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

Let's start by cleaning the column names so that the dataset is easier to work with.

We will use snakecase for the column names.

## 4. Cleaning Column Names

In [6]:
print("Original Column Names:")
print("----------------------")
print(autos.columns)


Original Column Names:
----------------------
Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
### Define a customized function to rename the column of this dataset.
### There are other more advanced technic to do this, but we will just use a simple basic function.

def rename_column(column_name):
    if column_name == "yearOfRegistration":
        new_name = "registration_year"
    elif column_name == "monthOfRegistration":
        new_name = "registration_month"   
    elif column_name == "notRepairedDamage":
        new_name = "unrepaired_damage" 
    elif column_name == "dateCreated":
        new_name = "ad_created" 
    elif column_name == "dateCrawled":
        new_name = "date_crawled" 
    elif column_name == "offerType":
        new_name = "offer_type" 
    elif column_name == "abtest":
        new_name = "ab_test" 
    elif column_name == "vehicleType":
        new_name = "vehicle_type" 
    elif column_name == "gearbox":
        new_name = "gear_box" 
    elif column_name == "powerPS":
        new_name = "power_PS" 
    elif column_name == "fuelType":
        new_name = "fuel_type" 
    elif column_name == "nrOfPictures":
        new_name = "no_of_pictures" 
    elif column_name == "postalCode":
        new_name = "postal_code" 
    elif column_name == "lastSeen":
        new_name = "last_seen" 
    else:
        new_name = column_name
    return new_name


### Looping through each column in the dataset and call the function to rename the column.

new_columns = []

for c in autos.columns:
    new_name = rename_column(c)
    new_columns.append(new_name)

autos.columns = new_columns


In [8]:
### Alternatively, can also directly assign the column names as shown below.

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen']

In [9]:
print("Renamed Column Names:")
print("---------------------")
print(autos.columns)


Renamed Column Names:
---------------------
Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [10]:
### Showing the first 5 rows of data with new column names.

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## 5. Initial Exploration - Checking General Statistics of the Dataset

In [11]:
### Check statistics for only numeric columns.

autos.describe()

registration_year      power_PS  registration_month  no_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [12]:
### Check statistics for object columns (non-numeric columns).

autos.describe(include=['O'])

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   

       vehicle_type gear_box  model   odometer fuel_type       brand  \
count         44905    47320  47242      50000     45518       50000   
unique            8        2    245         13         7          40   
top       limousine  manuell   golf  150,000km    benzin  volkswagen   
freq          12859    36993   4024      32424     30107       10687   

       unrepaired_damage           ad_created            last_seen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

In [13]:
### Check statistics for all columns.

autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

**Observation:**

Columns that have mostly one value (or where almost all of the values are the same) are candidates to be dropped:

- `seller` (almost all records are "privat")
- `offer_type` (almost all records are "Angebot")

The `no_of_pictures` column looks odd, we'll need to investigate this further.

In [14]:
### Further investigation on the 'no_of_pictures' column.
### All the values in this column is 0.

autos["no_of_pictures"].value_counts()

0    50000
Name: no_of_pictures, dtype: int64

## 6. Dropping Columns

Drop the following columns that have mostly one value (or where almost all of the values are the same):
- `seller` (almost all records are "privat")
- `offer_type` (almost all records are "Angebot")
- `no_of_pictures` (all records are "0")

In [15]:
autos = autos.drop(["seller", "offer_type", "no_of_pictures"], axis=1)

## 7. Cleaning Numeric Data Stored As Text

### 7.1 Converting the 'price' and 'odometer' column to numeric type

In [16]:
### Replace dollar sign and comma with empty string before converting the 'price' column to int type.

autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)


### Replace 'km' with empty string before converting the 'odometer' column to int type.

autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

### 7.2 Renaming the 'odometer' column to 'odometer_km'

In [17]:
### Renaming the 'odometer' column to 'odometer_km' so that we know the numeric values are in km.

autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

### Note: Either use inplace=True or assign the result back to the dataframe; otherwise, the modifications will be lost.


## 8. Identify Outliers (values that look unrealistically high or low)

### 8.1 Inspecting the 'price' Column

In [18]:
### Checking the number of unique price values.

autos["price"].unique().shape

(2357,)

In [19]:
### Checking statistics of the price column.

autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [20]:
### Checking the count for each unique price value.

autos["price"].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

In [21]:
autos["price"].value_counts(normalize=True).sort_index().head(10)


0     0.02842
1     0.00312
2     0.00006
3     0.00002
5     0.00004
8     0.00002
9     0.00002
10    0.00014
11    0.00004
12    0.00006
Name: price, dtype: float64

In [22]:
### Take a closer look on prices that are on the high end (sort by descending order).

autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

**Observation:**

- There are 2357 unique price values.

- 1421 records are with 0 price (which is about 2.8% of the listing), and the maximum price is one hundred million (99,999,999). 

- For prices that are on the high end, it seems like prices increased steadily until 350,000 and from there jump to unrealistically high values. There are 14 records with prices greater than 350,000.

- These unrealistic records, which are less than 3% of the listing, shall be removed later.

### 8.2 Inspecting the 'odometer_km' Column

In [23]:
### Checking the number of unique mileage values.

autos["odometer_km"].unique().shape

(13,)

In [24]:
### Checking statistics of the mileage.

autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [25]:
### Checking the count for each unique mileage value.

autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

**Observation:**

There seems to be more records with high mileage than low mileage in the listing.

### 8.3 Inspecting Dates

As seen below, dates for `date_crawled`, `ad_created`, and `last_seen` are all identified as string values by pandas.
We need to convert the data into a numerical representation so we can understand it quantitatively.


In [26]:
### Showing the first 5 records of 'date_crawled', 'ad_created', and 'last_seen'.

autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

### 8.3.1 Inspecting 'date_crawled'

In [27]:
### Select only the first 10 characters of the date to generate a distribution, and then sort by the index.
### To include missing values in the distribution with 'dropna=False'
### Use sort_index() to rank by date in ascending order.

autos["date_crawled"].str[:10].value_counts(dropna=False).sort_index()

# Following code is to use percentages instead of counts with 'normalize=True'.
# autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1665
2016-03-09    1661
2016-03-10    1606
2016-03-11    1624
2016-03-12    1839
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1576
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1876
2016-03-22    1647
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1709
2016-03-30    1681
2016-03-31    1596
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1826
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

In [28]:
autos["date_crawled"].str[:10].describe()

count          50000
unique            34
top       2016-04-03
freq            1934
Name: date_crawled, dtype: object

**Observation:**

- There are 34 unique date value for `date_crawled`.
- Looks like the website is crawled daily for about a month in March and April 2016 (from 2016-03-05 to 2016-04-07).
- The distribution of listings crawled on each day is roughly uniform.
- The date with the highest number of ads crawled, 1934, is on 2016-04-03.

### 8.3.2 Inspecting 'ad_created' Date

In [29]:
autos["ad_created"].str[:10].describe()

count          50000
unique            76
top       2016-04-03
freq            1946
Name: ad_created, dtype: object

In [30]:
### Very few ads are created on 2015.
### Let's check further below.

autos["ad_created"].str[:10].value_counts(dropna=False).sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
              ... 
2016-04-03    1946
2016-04-04    1844
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, Length: 76, dtype: int64

In [31]:
### Noticed that there are significant number of ads created on and after March 2016.
### Very few ads are created before March 2016.

autos["ad_created"].str[:10].value_counts(dropna=False).sort_index().head(50)

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       2
2016-02-11       1
2016-02-12       3
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
2016-02-22       1
2016-02-23       4
2016-02-24       2
2016-02-25       3
2016-02-26       2
2016-02-27       6
2016-02-28      10
2016-02-29       8
2016-03-01       5
2016-03-02       5
2016-03-03      43
2016-03-04      72
2016-03-05    1152
2016-03-06     756
2016-03-07    1737
2016-03-08    1667
2016-03-09    1662
2016-03-10    1593
2016-03-11    1639
2016-03-12    1831
Name: ad_created, dtype: int64

In [32]:
autos["ad_created"].str[:10].value_counts(dropna=False).sort_index().tail(50)

2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
2016-02-22       1
2016-02-23       4
2016-02-24       2
2016-02-25       3
2016-02-26       2
2016-02-27       6
2016-02-28      10
2016-02-29       8
2016-03-01       5
2016-03-02       5
2016-03-03      43
2016-03-04      72
2016-03-05    1152
2016-03-06     756
2016-03-07    1737
2016-03-08    1667
2016-03-09    1662
2016-03-10    1593
2016-03-11    1639
2016-03-12    1831
2016-03-13     846
2016-03-14    1761
2016-03-15    1687
2016-03-16    1500
2016-03-17    1560
2016-03-18     686
2016-03-19    1692
2016-03-20    1893
2016-03-21    1886
2016-03-22    1640
2016-03-23    1609
2016-03-24    1454
2016-03-25    1594
2016-03-26    1628
2016-03-27    1545
2016-03-28    1748
2016-03-29    1707
2016-03-30    1672
2016-03-31    1596
2016-04-01    1690
2016-04-02    1754
2016-04-03    1946
2016-04-04    1844
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, dtype: int64

**Observation:**

- There are 76 unique date value for `ad_created`.

- Noticed that there are significant number of ads created on and after March 2016, particularly during the crawling period which is from 2016-03-05 to 2016-04-07.

- Very few ads are created before March 2016.

- The date with the highest number of ads created, 1946, is on 2016-04-03.

- This is also the same date where the highest number of ads are crawled.

### 8.3.3 Inspecting 'last_seen' Date

In [33]:
autos["last_seen"].str[:10].describe()

count          50000
unique            34
top       2016-04-06
freq           11050
Name: last_seen, dtype: object

In [34]:
### There is significant number of ads that are last seen on 2016-04-06.

autos["last_seen"].str[:10].value_counts(dropna=False).sort_index()

2016-03-05       54
2016-03-06      221
2016-03-07      268
2016-03-08      380
2016-03-09      493
2016-03-10      538
2016-03-11      626
2016-03-12     1191
2016-03-13      449
2016-03-14      640
2016-03-15      794
2016-03-16      822
2016-03-17     1396
2016-03-18      371
2016-03-19      787
2016-03-20     1035
2016-03-21     1037
2016-03-22     1079
2016-03-23      929
2016-03-24      978
2016-03-25      960
2016-03-26      848
2016-03-27      801
2016-03-28     1043
2016-03-29     1117
2016-03-30     1242
2016-03-31     1192
2016-04-01     1155
2016-04-02     1245
2016-04-03     1268
2016-04-04     1231
2016-04-05     6214
2016-04-06    11050
2016-04-07     6546
Name: last_seen, dtype: int64

**Observation:**

- There are 34 unique date value for `last_seen`.

- There is a spike of records (6214, 11050, 6546) with `last seen` date on the last 3 days of the crawling period.

- The highest number of records (11050) are last seen on 2016-04-06, a day before the crawling period ends.

- This is unlikely due to a spike in sales.

### 8.4 Inspecting 'registration_year'

In [35]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [36]:
autos["registration_year"].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

In [37]:
autos["registration_year"].value_counts().sort_index()

1000    1
1001    1
1111    1
1500    1
1800    2
       ..
6200    1
8888    1
9000    2
9996    1
9999    4
Name: registration_year, Length: 97, dtype: int64

**Observation:**

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [38]:
### Checking the number of records where 'registration_year' is before 1900 OR after 2016.

autos[(autos["registration_year"] < 1900) | (autos["registration_year"] > 2016)]
   

date_crawled                                               name  \
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
65     2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
84     2016-03-27 19:52:54                                     Renault_twingo   
...                    ...                                                ...   
49796  2016-03-09 09:38:38                        Opel_corsa_1.4_zu_verkaufen   
49841  2016-03-11 15:37:02                                  Passat_abzugeben.   
49880  2016-03-30 08:52:57                                E39_528i_an_Bastler   
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   
49935  2016-04-01 21:48:20  Mercedes_A_klasse_angemeldet_mit_Tuef_und_Auto...   

       price  ab_test vehicle_type  registration_year   gear_box  power_PS  \
10       999     test          NaN               2017    manuell        90   
55         1     test          NaN               2017  automatik       224   
65       250  control          NaN               2017    manuell        65   
68     10990     test          NaN               2017    manuell       174   
84       900  control          NaN               2018        NaN        60   
...      ...      ...          ...                ...        ...       ...   
49796   4500     test          NaN               2017    manuell        90   
49841    600     test          NaN               2017    manuell       101   
49880      0  control          NaN               2017    manuell       193   
49910  22200     test          NaN               9000  automatik       140   
49935    800     test          NaN               2017  automatik       101   

          model  odometer_km  registration_month fuel_type          brand  \
10          NaN       150000                   4    benzin     volkswagen   
55     e_klasse       125000                   7    benzin  mercedes_benz   
65       fiesta       125000                   9    benzin           ford   
68      clubman       100000                   0       NaN           mini   
84       twingo       150000                   0       NaN        renault   
...         ...          ...                 ...       ...            ...   
49796     corsa        70000                   7    benzin           opel   
49841    passat       150000                   7       NaN     volkswagen   
49880       5er       150000                   4       NaN            bmw   
49910    andere        10000                   3    benzin           opel   
49935  a_klasse       150000                   9       NaN  mercedes_benz   

      unrepaired_damage           ad_created  postal_code            last_seen  
10                 nein  2016-03-14 00:00:00        86157  2016-04-07 03:16:21  
55                 nein  2016-03-06 00:00:00        22111  2016-03-08 05:45:44  
65                  NaN  2016-04-04 00:00:00        65606  2016-04-05 12:22:12  
68                 nein  2016-04-03 00:00:00        83135  2016-04-05 17:26:26  
84                  NaN  2016-03-27 00:00:00        40589  2016-04-05 18:46:49  
...                 ...                  ...          ...                  ...  
49796              nein  2016-03-09 00:00:00        88433  2016-03-17 20:45:08  
49841               NaN  2016-03-11 00:00:00        53804  2016-03-11 16:41:14  
49880                ja  2016-03-30 00:00:00        65468  2016-04-07 01:15:27  
49910               NaN  2016-04-03 00:00:00        56856  2016-04-05 22:18:26  
49935              nein  2016-04-01 00:00:00        39108  2016-04-01 21:48:20  

[1972 rows x 17 columns]

There are 1972 records where 'registration_year' is before 1900 OR after 2016, which is about 4%.

We will remove these outliers in the following section.

## 9. Remove Outliers

### 9.1 Remove records with unrealistic 'registration_year'

In [39]:
### Filter records where 'registration_year' is between 1900 and 2016.

boo = autos["registration_year"].between(1900, 2016)
autos2 = autos.loc[boo]
autos2.describe()

price  registration_year      power_PS    odometer_km  \
count  4.802800e+04        48028.00000  48028.000000   48028.000000   
mean   9.585252e+03         2002.80351    117.070417  125544.161739   
std    4.843817e+05            7.31085    195.151278   40106.751417   
min    0.000000e+00         1910.00000      0.000000    5000.000000   
25%    1.150000e+03         1999.00000     71.000000  100000.000000   
50%    2.990000e+03         2003.00000    107.000000  150000.000000   
75%    7.400000e+03         2008.00000    150.000000  150000.000000   
max    1.000000e+08         2016.00000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48028.000000  48028.000000  
mean             5.767760  50935.867327  
std              3.696802  25792.079828  
min              0.000000   1067.000000  
25%              3.000000  30459.000000  
50%              6.000000  49696.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

In [40]:
autos2["registration_year"].value_counts(normalize=True)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

### 9.2 Remove records with unrealistic 'price'

In [41]:
### Filter records where 'price' is not 0, and not greater than 350,000.

boo = autos2["price"].between(1, 350000)
autos2 = autos2.loc[boo]
autos2.describe()

price  registration_year      power_PS    odometer_km  \
count   46681.000000       46681.000000  46681.000000   46681.000000   
mean     5977.716801        2002.910756    117.892933  125586.855466   
std      9177.909479           7.185103    184.922911   39852.528628   
min         1.000000        1910.000000      0.000000    5000.000000   
25%      1250.000000        1999.000000     75.000000  100000.000000   
50%      3100.000000        2003.000000    109.000000  150000.000000   
75%      7500.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        2016.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        46681.000000  46681.000000  
mean             5.827125  51097.434181  
std              3.670300  25755.387192  
min              0.000000   1067.000000  
25%              3.000000  30827.000000  
50%              6.000000  49828.000000  
75%              9.000000  71732.000000  
max             12.000000  99998.000000

## 10. Find Popular Brands

In [42]:
### Create an array of unique brands

brands = autos2["brand"].unique()
print(brands)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


In [43]:
### Getting the count for each unique brand.

autos2["brand"].value_counts()

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

In [44]:
### Selecting the top 10 brands to be used for further analysis.

top10_brands = autos2["brand"].value_counts().head(10)
print(top10_brands)

volkswagen       9862
bmw              5137
opel             5022
mercedes_benz    4503
audi             4041
ford             3263
renault          2201
peugeot          1393
fiat             1197
seat              853
Name: brand, dtype: int64


In [45]:
### Getting the index label of the top 10 brands.

top10_brands = top10_brands.index
print(top10_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


**Observation:**

Volkswagen is the most popular brand, having 9862 records, which is almost double of it's first rival, BMW (5137 records).

## 11. Analyze the Top 10 Brands

In [46]:
### Create a general function to sort a given dictionary in descending order.
### This function will return a sorted list of tuple.

### Note that sorted() built-in function doesn't work too well with dictionaries because  
###      it only considers and returns the dictionary keys (instead of the key+value pair).
### We need to transform the dictionary into a list of tuples in order to do sorting.


def sort_dict(dictionary):
    ### Initialise a blank list to store tuples.
    list_of_tuple = []
    
    for key in dictionary:                      
        val_key_tuple = (dictionary[key], key)
        list_of_tuple.append(val_key_tuple)
        
    list_of_tuple_sorted = sorted(list_of_tuple, reverse = True)
    
    # Either print the sorted records, or return the sorted list of tuple.
    #for item in list_of_tuple_sorted:
    #    print(item[1], ':', item[0])
        
    return list_of_tuple_sorted
        

### 11.1 Exploring Price by Brand

In [47]:
### Getting the average price for each of the top 10 brands.

avg_price_by_brand = {}

for b in top10_brands:
    selected_rows = autos2[autos2["brand"] == b]
    mean = selected_rows["price"].mean()
    avg_price_by_brand[b] = int(mean)
    
print("### Average Price by Brand ###")
avg_price_by_brand


### Average Price by Brand ###


{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397}

In [48]:
### Call the sort_dict() function to sort the records.

print("### Average Price by Brand (Sorted) ###")
sort_dict(avg_price_by_brand)


### Average Price by Brand (Sorted) ###


[(9336, 'audi'),
 (8628, 'mercedes_benz'),
 (8332, 'bmw'),
 (5402, 'volkswagen'),
 (4397, 'seat'),
 (3749, 'ford'),
 (3094, 'peugeot'),
 (2975, 'opel'),
 (2813, 'fiat'),
 (2474, 'renault')]

**Observation:**

In the top 10 brands, there's a distinct price gap.

- Audi, Mercedes Benz, and BMW are more expensive.
- Opel, Fiat, and Renault are less expensive.
- Volkswagen, Seat, Ford and Peugeot are in between.

This may explain why Volkswagen is the most popular brand, as it's price is in between.

### 11.2 Exploring Mileage by Brand

In [49]:
### Getting the average mileage for each of the top 10 brands.

avg_mileage_by_brand = {}

for b in top10_brands:
    selected_rows = autos2[autos2["brand"] == b]
    mean = selected_rows["odometer_km"].mean()
    avg_mileage_by_brand[b] = int(mean)
    
#print("### Average Mileage by Brand ###")
#avg_mileage_by_brand

### Call the sort_dict() function to sort the records.

print("### Average Mileage by Brand (Sorted) ###")
sort_dict(avg_mileage_by_brand)


### Average Mileage by Brand (Sorted) ###


[(132572, 'bmw'),
 (130788, 'mercedes_benz'),
 (129310, 'opel'),
 (129157, 'audi'),
 (128707, 'volkswagen'),
 (128071, 'renault'),
 (127153, 'peugeot'),
 (124266, 'ford'),
 (121131, 'seat'),
 (117121, 'fiat')]

**Observation:**

All the top 10 brands have high mileage.

### 11.3 Analyzing Price & Mileage by Brand (Part 1)

Let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.
 
The following will combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly.

In [50]:
### Convert the 'avg_price_by_brand' dictionary to series objects, using the series constructor.
### The keys in the dictionary became the index in the series object.

bmp_series = pd.Series(avg_price_by_brand)
print(bmp_series)

### We can then create a single-column dataframe from this series object.
### We need to use the columns parameter when calling the dataframe constructor (which accepts a array-like object) 
###     to specify the column name (or the column name will be set to 0 by default).

df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
renault          2474
peugeot          3094
fiat             2813
seat             4397
dtype: int64


mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749
renault              2474
peugeot              3094
fiat                 2813
seat                 4397

In [51]:
### Adding 'mean_mileage' new column to the dataframe.

df["mean_mileage"] = pd.Series(avg_mileage_by_brand)
df

mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266
renault              2474        128071
peugeot              3094        127153
fiat                 2813        117121
seat                 4397        121131

**Observation:**

The average mileage seems to have no visible link with mean price.

There is no significant difference on the average mileage for cars with expensive or cheaper price.


### 11.4 Analyzing Price & Mileage by Brand (Part 2)

Split the mileage (odometer_km) into groups, and use aggregation to see if average prices follows any patterns based on the mileage.

In [52]:
### Checking the count and unique values of 'odometer_km'.

autos2["odometer_km"].value_counts().sort_index()

5000        785
10000       241
20000       742
30000       760
40000       797
50000       993
60000      1128
70000      1187
80000      1375
90000      1673
100000     2058
125000     4857
150000    30085
Name: odometer_km, dtype: int64

In [53]:
### Split the odometer_km into 3 groups.
### Getting the average price for each of the top 10 brands, for each of the 'odometer_km' groups.

avg_price_group1 = {}
avg_price_group2 = {}
avg_price_group3 = {}

for b in top10_brands:
    selected_rows_group1 = autos2[(autos2["brand"] == b) & (autos2["odometer_km"] <= 50000)]
    selected_rows_group2 = autos2[(autos2["brand"] == b) & (autos2["odometer_km"].between(50001, 100000))]
    selected_rows_group3 = autos2[(autos2["brand"] == b) & (autos2["odometer_km"] > 100000)]
    
    mean_group1 = selected_rows_group1["price"].mean()
    mean_group2 = selected_rows_group2["price"].mean()
    mean_group3 = selected_rows_group3["price"].mean()
    
    avg_price_group1[b] = int(mean_group1)
    avg_price_group2[b] = int(mean_group2)
    avg_price_group3[b] = int(mean_group3)
    

### Convert the 'avg_price_with_damage' dictionary to series objects, using the series constructor.
### The keys in the dictionary became the index in the series object.

mp_series = pd.Series(avg_price_group1)

### We can then create a single-column dataframe from this series object.
### We need to use the columns parameter when calling the dataframe constructor (which accepts a array-like object) 
###     to specify the column name (or the column name will be set to 0 by default).

df = pd.DataFrame(mp_series, columns=['mean_price_odo_up_to_50km'])

### Adding 'mean_price_non_damage' new column to the dataframe.

df["mean_price_odo_50_to_100km"] = pd.Series(avg_price_group2)
df["mean_price_odo_above_100km"] = pd.Series(avg_price_group3)
df


mean_price_odo_up_to_50km  mean_price_odo_50_to_100km  \
volkswagen                         13427                        9938   
bmw                                21404                       15799   
opel                                7980                        5389   
mercedes_benz                      22213                       16487   
audi                               25087                       17476   
ford                               10878                        5561   
renault                             7522                        4182   
peugeot                             7963                        5272   
fiat                                6626                        4060   
seat                               11137                        6849   

               mean_price_odo_above_100km  
volkswagen                           3762  
bmw                                  6282  
opel                                 2099  
mercedes_benz                        6164  
audi                                 6303  
ford                                 2440  
renault                              1659  
peugeot                              2139  
fiat                                 1758  
seat                                 2656

**Observation:**

For the same brand, average price is reduced when the mileage is increased.

### 11.5 Analyzing Price For Cars With Damaged and Non-Damage

Here we would want to find out how much cheaper are cars with damage than their non-damaged counterparts.

In [54]:
### Checking the unique values for the 'unrepaired_damage' column.

autos2["unrepaired_damage"].value_counts()

nein    33834
ja       4540
Name: unrepaired_damage, dtype: int64

In [55]:
### Getting the average price for each of the top 10 brands, for cars with damage and without damage.

avg_price_with_damage = {}
avg_price_non_damaged = {}

for b in top10_brands:
    selected_rows_with_damage = autos2[(autos2["brand"] == b) & (autos2["unrepaired_damage"] == "ja")]
    selected_rows_non_damage = autos2[(autos2["brand"] == b) & (autos2["unrepaired_damage"] == "nein")]
    
    mean_with_damage = selected_rows_with_damage["price"].mean()
    mean_non_damage = selected_rows_non_damage["price"].mean()
    
    avg_price_with_damage[b] = int(mean_with_damage)
    avg_price_non_damaged[b] = int(mean_non_damage)
    

### Convert the 'avg_price_with_damage' dictionary to series objects, using the series constructor.
### The keys in the dictionary became the index in the series object.

mp_series = pd.Series(avg_price_with_damage)

### We can then create a single-column dataframe from this series object.
### We need to use the columns parameter when calling the dataframe constructor (which accepts a array-like object) 
###     to specify the column name (or the column name will be set to 0 by default).

df = pd.DataFrame(mp_series, columns=['mean_price_with_damage'])

### Adding 'mean_price_non_damage' new column to the dataframe.

df["mean_price_non_damage"] = pd.Series(avg_price_non_damaged)
df


mean_price_with_damage  mean_price_non_damage
volkswagen                       2179                   6469
bmw                              3512                   9437
opel                             1367                   3660
mercedes_benz                    3921                   9798
audi                             3324                  10914
ford                             1375                   4660
renault                          1145                   3103
peugeot                          1372                   3674
fiat                             1146                   3446
seat                             1741                   5183

In [56]:
### Calculate the percentage and adding it as column to the dataframe.

df["percent_cheaper"] = (df["mean_price_with_damage"] / df["mean_price_non_damage"]) * 100
df

mean_price_with_damage  mean_price_non_damage  percent_cheaper
volkswagen                       2179                   6469        33.683722
bmw                              3512                   9437        37.215217
opel                             1367                   3660        37.349727
mercedes_benz                    3921                   9798        40.018371
audi                             3324                  10914        30.456295
ford                             1375                   4660        29.506438
renault                          1145                   3103        36.899774
peugeot                          1372                   3674        37.343495
fiat                             1146                   3446        33.255949
seat                             1741                   5183        33.590585

**Observation:**

For the top 10 brands, cars with damage are around 30-40% cheaper than their non-damaged counterparts.

## 12. Find the most common brand/model combinations

In [57]:
### Create a new column 'brand_and_model', which combines the brand and the model columns.

autos2["brand_and_model"] = autos2["brand"] + "/" + autos2["model"]
autos2

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehicle_type  registration_year   gear_box  power_PS  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code  \
0                  nein  2016-03-26 00:00:00        79588   
1                  nein  2016-04-04 00:00:00        71034   
2                  nein  2016-03-26 00:00:00        35394   
3                  nein  2016-03-12 00:00:00        33729   
4                  nein  2016-04-01 00:00:00        39218   
...                 ...                  ...          ...   
49995              nein  2016-03-27 00:00:00        82131   
49996              nein  2016-03-28 00:00:00        44807   
49997              nein  2016-04-02 00:00:00        73430   
49998              nein  2016-03-08 00:00:00        35683   
49999              nein  2016-03-13 00:00:00        45897   

                 last_seen  brand_and_model  
0      2016-04-06 06:45:54   peugeot/andere  
1      2016-04-06 14:45:08          bmw/7er  
2      2016-04-06 20:15:37  volkswagen/golf  
3      2016-03-15 03:16:28     smart/fortwo  
4      2016-04-01 14:38:50       ford/focus  
...                    ...              ...  
49995  2016-04-01 13:47:40          audi/q5  
49996  2016-04-02 14:18:02       opel/astra  
49997  2016-04-04 11:47:27         fiat/500  
49998  2016-04-05 16:45:07          audi/a3  


In [58]:
### Getting the count for each unique brand/model combination.

autos2["brand_and_model"].value_counts()

volkswagen/golf      3707
bmw/3er              2615
volkswagen/polo      1609
opel/corsa           1592
volkswagen/passat    1349
                     ... 
ford/b_max              1
rover/discovery         1
rover/rangerover        1
bmw/i3                  1
audi/200                1
Name: brand_and_model, Length: 290, dtype: int64

**Observation:**

The most common brand/model combination is Volkswagen/Golf (3707 records), followed by BMW/3er (2615 records).